### PR TITLE
Automatically publish artifacts to Maven Central repository

### LINKED ISSUE #2264: Upload snapshot builds to `oss.sonatype.org`
Our users sometimes wishes to try the SNAPSHOT builds to preview the upcoming changes.

I need to figure out how to pass the `oss.sonatype.org` credencials safely, e.g. do *not* pass them when building a pull request and do *not* include them in build output such as console output and report files.

### PR DESCRIPTION
**Pull request by ikhoon**
Motivation:

It takes 20 to 30 minutes to upload new artifacts to Maven Central
repository that might not be good for the mental health of devs. 🤦

Modifications:

- Use `io.github.gradle-nexus.publish-plugin` to automatically close
  and release a staging repository in Sonatype.
- Add new GitHub Actions jobs to publish new artifacts when a new
  tag is created or a new commit is pushed to master branch.
- Use environment variables to set GPG credentials.
- Miscellaneous
  - Make GitHub Actions jobs run only on `line/armeria` repository.

Result:

- Users can early access a snapshot version when a PR is merged.
- Simplify release processes.
- Closes #2264

### CODE REVIEW

File: .github/workflows/publish-release.yml [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+            ~/.gradle/wrapper
+            ~/.gradle/caches
+          key: build-${{ matrix.java }}-${{ runner.os }}-${{ secrets.CACHE_VERSION }}-${{ hashFiles('gradle.properties', 'gradle/wrapper/gradle-wrapper.properties', '**/build.gradle', 'dependencies.yml', '*/package-lock.json') }}
+          restore-keys: |
+            build-${{ matrix.java }}-${{ runner.os }}-${{ secrets.CACHE_VERSION }}-
+            build-${{ matrix.java }}-${{ runner.os }}-
+
+      - name: Publish and close repository
+        run: |
+          ./gradlew --no-daemon --stacktrace --max-workers=1 publish closeAndReleaseStagingRepository
```
Comment (anuraaga) [inline .github/workflows/publish-release.yml:36]: Does publish run the tests also? I think normally you would run `build` as a step before publish, unless you're confident in 100% reproducibility of build (guessing not)

Comment (ikhoon) [reply .github/workflows/publish-release.yml:36]: > Does publish run the tests also? 

`publish` does not depend on `test` task. There is a test job that is triggered when a new commit is pushed to the master branch. 
https://github.com/line/armeria/blob/e294957e2999d361c3a2450ac85a2b54c1d2322b/.github/workflows/actions_build.yml#L3-L5
I thought it was enough for verification. But it makes sense to run `build` for double check.  👍

Comment (minwoox) [reply .github/workflows/publish-release.yml]: >  But it makes sense to run build for double check. 👍

Then, should we probably publish the artifact when the build succeeds?

Comment (ikhoon) [reply .github/workflows/publish-release.yml]: Yes, it is. It is noted that flaky tests are excluded. I think the build should pass in the self-hosted runners.

**Review by minwoox — APPROVED** [review]
Let me give you  👍 * 100.

File: .github/workflows/publish-snapshot.yml [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+            build-${{ matrix.java }}-${{ runner.os }}-${{ secrets.CACHE_VERSION }}-
+            build-${{ matrix.java }}-${{ runner.os }}-
+
+      - name: Publish snapshots
+        run: |
+          ./gradlew --no-daemon --stacktrace --max-workers=1 publish
+        env:
+          # Should not use '-P' option with 'secrets' that can cause unexpected results
+          # if secret values contains white spaces or new lines.
+          ORG_GRADLE_PROJECT_ossrhUsername: ${{ secrets.OSSRH_USER_NAME }}
```
[the remark this answers is not in the document]
Comment (anuraaga) [reply .github/workflows/publish-snapshot.yml:46]: By the way don't you want to only publish snapshots if main build succeeds (so should be step in main build workflow)?

[the remark this answers is not in the document]
Comment (minwoox) [reply .github/workflows/publish-snapshot.yml:46]: @ikhoon Could you address this, please? 😄

[the remark this answers is not in the document]
Comment (ikhoon) [reply .github/workflows/publish-snapshot.yml:46]: Sorry for the huge delay. Let me address the valuable comments. Thanks, @anuraaga! 🙇‍♂️

File: .post-release-msg [hunk 3]
```diff
@@ -1,17 +1,13 @@
-1. Upload the artifacts to the staging repository and prepare the web site:
+1. Build the web site:
 
    git checkout ${tag} && \
-   ./gradlew --no-daemon clean publish site
+   ./gradlew --no-daemon clean site
 
-2. Close and release the staging repository at:
-
-   https://oss.sonatype.org/
```
Comment (trustin) [inline .post-release-msg:8]: Maybe add a step that tells the dev to wait until the artifacts are up in the Central?

**Review by trustin — APPROVED** [review]
LGTM once @anuraaga's comment is resolved.

File: .github/workflows/publish-release.yml [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+            build-${{ matrix.java }}-${{ runner.os }}-
+
+      - name: Build with Gradle
+        run: |
+          ./gradlew --no-daemon --stacktrace --max-workers=8 --parallel -PflakyTests=false build
+        shell: bash
+
+      - name: Publish and close repository
+        run: |
+          ./gradlew --no-daemon --stacktrace --max-workers=1 clean publish closeAndReleaseStagingRepository
```
Comment (anuraaga) [inline .github/workflows/publish-release.yml:41]:
```suggestion
          ./gradlew --no-daemon --stacktrace --max-workers=1 publish closeAndReleaseStagingRepository
```

Don't think it's common to clean+rebuild the tested artifacts before publishing them

**Comment by ikhoon** [comment]
Sent a PR for the changes in `common-publish.gradle` https://github.com/line/gradle-scripts/pull/111

**Comment by trustin** [comment]
@ikhoon wrote:
> Sent a PR for the changes in `common-publish.gradle` https://github.com/line/gradle-scripts/pull/111

The PR has been merged and `master` has been updated with it.

**Review by jrhee17 — APPROVED** [review]
Awesome work! 👍 🙇 🥇 Thank you! @ikhoon !

**Review by trustin — APPROVED** [review]

**Pull request merged**